### ISSUE TITLE
`sops` can't decrypt when only provided an SSH key path

### ISSUE THREAD

**Issue by 3ulalia** [Mic92/sops-nix#695]
I've managed to pare my config down to the following minimal setup:

```nix
sops.defaultSopsFile = inputs.self.outPath + "/secrets/secrets.yaml";
sops.age.sshKeyPaths = ["/home/eulalia/.ssh/id_ed25519"];
sops.secrets."passwords/eulalia".neededForUsers = true;
```
Then, in my secrets file (~/flake/secrets/secrets.yaml):
```yaml
passwords:
  eulalia: $y$thisisafakepasswordhash
```
And in my sops yaml file (~/.sops.yaml):
```yaml
keys:
  - &eulalia thepublickeytomySSHkeywhenconvertedtoage
creation_rules:
  - path_regex: secrets/[^/]+\.(yaml|json|env|ini|sops)$
    key_groups:
      - age:
        - *eulalia
```
Despite this, when I run `sops ~/flake/secrets/secrets.yaml` it tells me that it is unable to decrypt:

```
eulalia@sunlanii:~/flake/ > sops secrets/secrets.yaml
Failed to get the data key required to decrypt the SOPS file.

Group 0: FAILED
  thepublickeytomySSHkeywhenconvertedtoage: FAILED
    - | failed to load age identities: failed to open file: open
      | /home/eulalia/.config/sops/age/keys.txt: no such file or
      | directory
```
The only way to get it to decrypt is to use ssh-to-age to convert my SSH key to an age key and then put it in that location.

Why do I have to put my key there? Isn't setting `sops.age.sshKeyPaths` enough?

(And yes, I have run `sops updatekeys ~/flake/secrets/secrets.yaml` a number of times. It has not fixed the issue :p)

I'm presuming that this is the reason behind a related issue I have - namely, when I set my password as described above, it works until I reboot, at which point I'm unable to login. I presume this is because `sops` is looking for a key file in `~/.config/sops/age/keys.txt` despite it not being there.

Thanks!

**Comment by n8henrie** [comment]
@3ulalia would you mind elaborating?

I've used agenix for a while and want to experiment with sops, but I'm running into the same misunderstanding.

Even if I specify `SOPS_AGE_SSH_PRIVATE_KEY_FILE=/etc/ssh/ssh_host_ed25519_key` [^0], I get `Failed to get the data key required to decrypt the SOPS file.` as it tries to read from `~/.config/sops/age/keys.txt` (which I've removed).

Why is `~/.config/sops/age/keys.txt` necessary when I have `/etc/ssh/ssh_host_ed25519_key`?

[^0]: https://github.com/getsops/sops?tab=readme-ov-file#23encrypting-using-age

**Comment by n8henrie** [comment]
Yes, I was just reading the docs for main which is still unreleased.

After building from source, I was able to add my ssh public key like so (initially keeping the age key version around):

```yaml
keys:
  - &myagekey age1234
  - &samekeyasssh ssh-ed25519 AAA...
creation_rules:
  - path_regex: secrets/[^/]+\.json$
    key_groups:
    - age:
      - *myagekey
      - *samekeyasssh
```

then update the keys so that both the ssh and age keys were recognized:

(I'm not sure what part of my env was required, bit failed without `-E`)

```console
$ sudo -E \
    SOPS_AGE_SSH_PRIVATE_KEY_FILE=/etc/ssh/ssh_host_ed25519_key \
    ~/go/src/github.com/getsops/sops/cmd/sops/sops \
    updatekeys secrets/sops.json
```

After which I was able to blow away the age directory and still decrypt / edit the test file:

```console
$ rm -r ~/.config/sops
$ sudo \
    SOPS_AGE_SSH_PRIVATE_KEY_FILE=/etc/ssh/ssh_host_ed25519_key \
    ~/go/src/github.com/getsops/sops/cmd/sops/sops \
    --decrypt \
    secrets/sops.json
{
        "hello": "Welcome to SOPS! Edit this file as you please!",
        "example_key": "example_value",
        "example_array": [
                "example_value1",
                "example_value2"
        ],
        "example_number": 1234.56789,
        "example_booleans": [
                true,
                false
        ]
}
```

And finally was able to remove the `age` version of the key from my sops config and once again `updatekeys`.

**Comment by travisty-** [comment]
Bumping this as SOPS officially added support for SSH keys in [v3.10.0](https://github.com/getsops/sops/releases/tag/v3.10.0).

I've been using only SSH keys in my own config for a couple months now and have had no issues, however they still get converted and written to `age-keys.txt`. I think this code path could be modified to remove the creation of those files in favor of letting SOPS handle it internally?

https://github.com/Mic92/sops-nix/blob/2c8def626f54708a9c38a5861866660395bb3461/pkgs/sops-install-secrets/main.go#L1328C1-L1361C3

The setup instructions could be simplified a bit at that point, since `ssh-to-age` would no longer be necessary.

**Comment by travisty-** [comment]
@nikhilmaddirala I added sops-nix to my config [in this commit](https://github.com/travisty-/dotfiles/commit/efe8bbc598d5a66a687c1e59b26e7bb922c38644) if you want to take a look. I started with SSH keys up front though, so I didn't have to update the keys for any existing secrets (as mentioned in the previous comments).
